File: csharp/killer-sudoku-helper/KillerSudokuHelper.cs
public static class KillerSudokuHelper
{
    public static IEnumerable<int[]> Combinations(int sum, int size, int[] exclude)
    {
        List<int[]> result = [];
        var excluded = exclude.ToHashSet();

        Generate(
            start: 1,
            remainingSum: sum,
            remainingSize: size,
            current: [],
            excluded: excluded,
            result: result
        );

        return result;
    }

    private static void Generate(int start, int remainingSum, int remainingSize, List<int> current, HashSet<int> excluded, List<int[]> result)
    {
        if (remainingSize == 0)
        {
            if (remainingSum == 0)
            {
                result.Add([.. current]);
            }

            return;
        }

        for (var digit = start; digit <= 9; digit++)
        {
            if (excluded.Contains(digit))
            {
                continue;
            }

            if (digit > remainingSum)
            {
                break;
            }

            current.Add(digit);

            Generate(
                start: digit + 1,
                remainingSum: remainingSum - digit,
                remainingSize: remainingSize - 1,
                current: current,
                excluded: excluded,
                result: result
            );

            current.RemoveAt(current.Count - 1);
        }
    }
}

File: java/killer-sudoku-helper/src/main/java/KillerSudokuHelper.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

public class KillerSudokuHelper {

    List<List<Integer>> combinationsInCage(Integer cageSum, Integer cageSize, List<Integer> exclude) {
        final var result = new ArrayList<ArrayList<Integer>>();

        generate(1, cageSum, cageSize, new ArrayList<>(), new HashSet<>(exclude), result);

        return List.copyOf(result);
    }

    List<List<Integer>> combinationsInCage(Integer cageSum, Integer cageSize) {
        return combinationsInCage(cageSum, cageSize, Collections.emptyList());
    }

    private static void generate(Integer start, Integer remainingSum, Integer remainingSize, ArrayList<Integer> current, HashSet<Integer> excluded, ArrayList<ArrayList<Integer>> result) {
        if (remainingSize == 0) {
            if (remainingSum == 0) {
                result.add(new ArrayList<>(current));
            }

            return;
        }

        for (var digit = start; digit <= 9; digit++) {
            if (excluded.contains(digit)) {
                continue;
            }

            if (digit > remainingSum) {
                break;
            }

            current.add(digit);

            generate(
                    digit + 1,
                    remainingSum - digit,
                    remainingSize - 1,
                    current,
                    excluded,
                    result
            );

            current.removeLast();
        }
    }

}
